# 🎞️ **MovieLens: что рассказывают 100 000 оценок о фильмах и людях** 🎞️

Датасет **ml-latest-small** от GroupLens содержит:

- 9 742 фильма
- 100 836 оценок
- 3 683 тега
- от 610 пользователей (1996–2018 гг.)

*В этом отчёте используются только классы и методы из movielens_analysis.py. Каждая ячейка сопровождается замером времени выполнения через %timeit, чтобы заодно оценить производительность каждого метода.*

## **I. Каталог фильмов: что вообще есть в базе?**

### 🎥 *Фильмы в датасете покрывают выпуск с 1902 по 2018 год*

- Основная масса сосредоточена в **1995–2018**
- **Пиковый год** по количеству добавленных фильмов - **2002** (310 фильмов), следом 2006 (295) и 2001 (294) 

In [ ]:
from movielens_analysis import Movies, Tags, Ratings, Links

MOVIES_PATH = "../datasets/movies.csv"
LINKS_PATH = "../datasets/links.csv"
TAGS_PATH = "../datasets/tags.csv"
RATINGS_PATH = "../datasets/ratings.csv"
TMDB_API_KEY = "API_KEY" 

%timeit Movies(MOVIES_PATH)
movies = Movies(MOVIES_PATH)

28.8 ms ± 7.18 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [18]:
%timeit movies.dist_by_release()
movies.dist_by_release()

594 μs ± 14.8 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


OrderedDict([(2002, 310),
             (2006, 295),
             (2001, 294),
             (2007, 284),
             (2000, 283),
             (2009, 282),
             (2003, 279),
             (2004, 279),
             (2014, 278),
             (1996, 276),
             (2015, 274),
             (2005, 273),
             (2008, 269),
             (1999, 263),
             (1997, 260),
             (1995, 259),
             (1998, 258),
             (2011, 254),
             (2010, 247),
             (2013, 239),
             (1994, 237),
             (2012, 233),
             (2016, 218),
             (1993, 198),
             (1992, 167),
             (1988, 165),
             (1987, 153),
             (1990, 147),
             (1991, 147),
             (2017, 147),
             (1989, 142),
             (1986, 139),
             (1985, 126),
             (1984, 101),
             (1981, 92),
             (1980, 89),
             (1982, 87),
             (1983, 83),
             (19

### 🎥 *Наиболее популярный жанр - драма*

**Драма** - 4 361 фильм (44,8% каталога) и **Комедия** - 3 756 (38,6%) доминируют, часто пересекаясь друг с другом и с другими жанрами (у фильма может быть несколько жанров одновременно, поэтому сумма процентов больше 100%). На другом полюсе - **Фильм-нуар** (87 фильмов, 0,9%) и **Вестерн** (167): нишевые жанры, которые почти вымерли в современном кинопроизводстве.

In [19]:
%timeit movies.dist_by_genres()
movies.dist_by_genres()

1.41 ms ± 108 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


{'Drama': 4361,
 'Comedy': 3756,
 'Thriller': 1894,
 'Action': 1828,
 'Romance': 1596,
 'Adventure': 1263,
 'Crime': 1199,
 'Sci-Fi': 980,
 'Horror': 978,
 'Fantasy': 779,
 'Children': 664,
 'Animation': 611,
 'Mystery': 573,
 'Documentary': 440,
 'War': 382,
 'Musical': 334,
 'Western': 167,
 'IMAX': 158,
 'Film-Noir': 87}

### 🎥 *Рекордсмен по числу жанров — «Rubber» (2010)*
Фильм Квентина Дюпье с 10 жанрами одновременно: сюжет рассказывает об автомобильной покрышке по имени Роберт, которая оживает в пустыне и начинает убивать людей и животных с помощью телекинеза

<img src="attachment:cc9113ce-8629-4b96-9279-83ea07cb7a83.png" width="150"> <img src="attachment:90aa72e2-4e9c-4c82-9c95-b323ee9a258b.png" width="150"> 

In [20]:
%timeit movies.most_genres(5)
movies.most_genres(5)

1.93 ms ± 76.1 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


{'Rubber (2010)': 10,
 'Patlabor: The Movie (Kidô keisatsu patorebâ: The Movie) (1989)': 8,
 'Mulan (1998)': 7,
 'Who Framed Roger Rabbit? (1988)': 7,
 'Osmosis Jones (2001)': 7}

## **II. Тэги: что пользователи говорят о фильмах**

### 💭 *"Это было мелодраматично и... немного глупо"*

**Самые многословные тэги - это не тэги в привычном понимании, а целые мини-рецензии**: 16-словное «Something for everyone in this one... saw it without and plan on seeing it with kids!» или резкое 10-словное высказывание о католической церкви. Пользователи MovieLens явно используют поле тэга как площадку для короткого мнения, а не только для ключевых слов

**Пересечение «самых многословных» и «самых длинных по символам» тэгов** - это одни и те же 3 тэга-рецензии. Логично: длинные фразы почти всегда содержат много слов, так что топы сильно перекрываются

In [21]:
%timeit Tags(TAGS_PATH)
tags = Tags(TAGS_PATH)

7.98 ms ± 1.32 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [22]:
%timeit tags.most_words(5)
tags.most_words(5)

927 μs ± 9.96 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


{'Something for everyone in this one... saw it without and plan on seeing it with kids!': 16,
 'the catholic church is the most corrupt organization in history': 10,
 'villain nonexistent or not needed for good story': 8,
 'It was melodramatic and kind of dumb': 7,
 'stop using useless characters for filler': 6}

In [23]:
%timeit tags.longest(5)
tags.longest(5)

227 μs ± 5.71 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


['Something for everyone in this one... saw it without and plan on seeing it with kids!',
 'the catholic church is the most corrupt organization in history',
 'villain nonexistent or not needed for good story',
 'r:disturbing violent content including rape',
 '06 Oscar Nominated Best Movie - Animation']

In [24]:
%timeit tags.most_words_and_longest(5)
tags.most_words_and_longest(5)

1.16 ms ± 9.09 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


['Something for everyone in this one... saw it without and plan on seeing it with kids!',
 'the catholic church is the most corrupt organization in history',
 'villain nonexistent or not needed for good story']

### 💭 Самый популярный тэг - «In Netflix queue» 

**А вот самый популярный тэг по числу применений** - вовсе не описание фильма, а «In Netflix queue» (131 раз): пользователи явно использовали теги MovieLens как личный to-watch список, а не только как способ описать сюжет. Из содержательных тегов лидируют «atmospheric» (36), «superhero» и «thought-provoking» (24), «funny» (23)

**Поиск по слову «funny» показывает, что теги живут своей неструктурированной жизнью:** 'Funny', 'funny', 'very funny', 'stupid but funny' и даже противоположность - 'not funny'. Простой поиск по подстроке без анализа тональности легко перепутал бы последний тег с положительной оценкой.

In [25]:
%timeit tags.most_popular(10)
tags.most_popular(10)

232 μs ± 6.69 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


{'In Netflix queue': 131,
 'atmospheric': 36,
 'superhero': 24,
 'thought-provoking': 24,
 'funny': 23,
 'Disney': 23,
 'surreal': 23,
 'religion': 22,
 'sci-fi': 21,
 'dark comedy': 21}

In [26]:
%timeit tags.tags_with("funny")
tags.tags_with("funny")

287 μs ± 5.14 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


['Funny', 'funny', 'not funny', 'stupid but funny', 'very funny']

## **III. Рейтинг: как пользователи оценивают фильмы?**

### ⭐ *Активность по годам неровная: резкий пик оценок в 2000 году*

**Резкий пик в 2000 году - 10 061 оценка, при том что соседние 1999 и 2001 годы дают лишь 2 439 и 3 922**. Это больше похоже на разовую пакетную загрузку исторических оценок (или технический артефакт при формировании датасета), чем на органический всплеск интереса пользователей. Более поздние пики (2007, 2015–2017) уже выглядят правдоподобнее и, вероятно, связаны с ростом самого сервиса и мобильных клиентов.

In [27]:
%timeit Ratings(RATINGS_PATH, path_to_movies=MOVIES_PATH)
ratings = Ratings(RATINGS_PATH, path_to_movies=MOVIES_PATH)
ratings_movies = ratings.Movies(ratings)

160 ms ± 62.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [28]:
%timeit ratings_movies.dist_by_year()
ratings_movies.dist_by_year()

32.5 ms ± 591 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)


{1996: 6040,
 1997: 1916,
 1998: 507,
 1999: 2439,
 2000: 10061,
 2001: 3922,
 2002: 3478,
 2003: 4014,
 2004: 3279,
 2005: 5813,
 2006: 4059,
 2007: 7114,
 2008: 4351,
 2009: 4158,
 2010: 2300,
 2011: 1690,
 2012: 4657,
 2013: 1664,
 2014: 1439,
 2015: 6616,
 2016: 6702,
 2017: 8199,
 2018: 6418}

### ⭐ *Люди в целом гораздо охотнее хвалят фильмы, чем ставят им единицу или того хуже*

- **Зрители явно предпочитают целые баллы половинчатым:** суммарно 69,9% всех оценок — это 1.0/2.0/3.0/4.0/5.0, и только 30,1% являются оценки вроде 3.5 или 4.5
- **Самая частая оценка** - 4.0 (26 818 раз, 26,6%), за ней 3.0 (20 047)
- **Реже всего ставят низший балл** - 0.5 (1 370 раз)

In [29]:
%timeit ratings_movies.dist_by_rating()
ratings_movies.dist_by_rating()

7.62 ms ± 310 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


{0.5: 1370,
 1.0: 2811,
 1.5: 1791,
 2.0: 7551,
 2.5: 5550,
 3.0: 20047,
 3.5: 13136,
 4.0: 26818,
 4.5: 8551,
 5.0: 13211}

### ⭐ *Топ по количеству оценок - Forrest Gump*

Далее по популярности оценивания идут: 
- Shawshank Redemption (317)
- Pulp Fiction (307)
- Silence of the Lambs (279)
- The Matrix (278)

**Это не лучшие фильмы по мнению зрителей, а самые охваченные - те, что видело больше всего людей из выборки**

<img src="attachment:f6f9d98c-c5ab-461a-b724-7de07cc267a2.png" width="150"> <img src="attachment:f660d307-0a8d-4ed3-8d30-41a4d84427b6.png" width="400"> 

### ⭐ *Топ по средней оценке - это классическая ловушка малых выборок*

У всей пятёрки лидеров (Jinx, Galaxy of Terror, Alien Contamination, «I'm the One That I Want», Lesson Faust) всего 1–2 оценки. Одна пятёрка от единственного зрителя автоматически даёт средний балл 5.0. Медианный топ почти такой же, но добавляет культовый «Troll 2» (3 оценки) - фильм, знаменитый именно тем, что считается «лучшим худшим фильмом всех времён»

Похожая ловушка и с «самыми спорными» фильмами по дисперсии оценок: у лидеров («Иваново детство» и «Fanny and Alexander», дисперсия 10.12) - **всего по 2 оценки, и это ровно 0.5 и 5.0**. Формально это максимальная возможная поляризация, но по сути это случайное совпадение двух полярных мнений при крошечной выборке. Интерпретировать результат стоит с поправкой на количество оценок.

In [30]:
%timeit ratings_movies.top_by_num_of_ratings(5)
ratings_movies.top_by_num_of_ratings(5)

3.09 ms ± 165 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


{'Forrest Gump (1994)': 329,
 'Shawshank Redemption, The (1994)': 317,
 'Pulp Fiction (1994)': 307,
 'Silence of the Lambs, The (1991)': 279,
 'Matrix, The (1999)': 278}

In [31]:
%timeit ratings_movies.top_by_ratings(5, metric="average")
ratings_movies.top_by_ratings(5, metric="average")

9.92 ms ± 194 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


{'The Jinx: The Life and Deaths of Robert Durst (2015)': 5.0,
 'Galaxy of Terror (Quest) (1981)': 5.0,
 'Alien Contamination (1980)': 5.0,
 "I'm the One That I Want (2000)": 5.0,
 'Lesson Faust (1994)': 5.0}

In [32]:
%timeit ratings_movies.top_controversial(5)
ratings_movies.top_controversial(5)

14.8 ms ± 345 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


{"Ivan's Childhood (a.k.a. My Name is Ivan) (Ivanovo detstvo) (1962)": 10.12,
 'Fanny and Alexander (Fanny och Alexander) (1982)': 10.12,
 'Lassie (1994)': 8.0,
 'Zed & Two Noughts, A (1985)': 8.0,
 'Kwaidan (Kaidan) (1964)': 8.0}

## **IV. Пользователи: кто оценивает фильмы?**

### 💁 Активность пользователей крайне неравномерна

**Cамый активный пользователь 414 поставил 2 698 оценок**, второй (599) - 2 478, а всего в базе 610 пользователей. По условиям датасета **у каждого минимум 20 оценок** - это видно и по «хвосту» распределения, где многие пользователи находятся ровно на этой границе

In [33]:
%timeit ratings.Users(ratings)
ratings_users = ratings.Users(ratings)

98.9 ns ± 0.58 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)


In [34]:
%timeit ratings_users.dist_by_num_of_ratings()
ratings_users.dist_by_num_of_ratings()

89.1 μs ± 977 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


{414: 2698,
 599: 2478,
 474: 2108,
 448: 1864,
 274: 1346,
 610: 1302,
 68: 1260,
 380: 1218,
 606: 1115,
 288: 1055,
 249: 1046,
 387: 1027,
 182: 977,
 307: 975,
 603: 943,
 298: 939,
 177: 904,
 318: 879,
 232: 862,
 480: 836,
 608: 831,
 600: 763,
 483: 728,
 590: 728,
 105: 722,
 19: 703,
 305: 677,
 489: 648,
 111: 646,
 438: 635,
 217: 613,
 140: 608,
 477: 600,
 555: 578,
 91: 575,
 28: 570,
 219: 528,
 534: 520,
 89: 518,
 64: 517,
 226: 507,
 561: 505,
 18: 502,
 525: 500,
 57: 476,
 381: 474,
 368: 469,
 509: 467,
 469: 465,
 560: 458,
 462: 455,
 292: 446,
 21: 443,
 597: 443,
 42: 440,
 160: 437,
 294: 437,
 580: 436,
 596: 411,
 202: 403,
 275: 403,
 517: 400,
 45: 399,
 156: 398,
 514: 397,
 391: 386,
 567: 385,
 357: 383,
 103: 377,
 339: 371,
 62: 366,
 199: 363,
 125: 360,
 51: 359,
 132: 347,
 66: 345,
 313: 340,
 200: 334,
 221: 331,
 6: 314,
 453: 311,
 50: 310,
 425: 306,
 428: 300,
 573: 299,
 352: 294,
 84: 293,
 122: 292,
 382: 291,
 356: 289,
 135: 279,
 239:

### 💁 Пользователь с самой высокой средней оценкой (5.0) поставил 20 оценок

И, судя по всему, оценивал только то, что действительно понравилось. Топ таких «добрых» пользователей (251, 515, 25, 30) держится в диапазоне 20-34 оценки - небольшая, заведомо позитивно смещённая выборка каждого.

In [35]:
%timeit ratings_users.dist_by_user_ratings(metric="average")
ratings_users.dist_by_user_ratings(metric="average")

1.22 ms ± 13.8 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


{53: 5.0,
 251: 4.87,
 515: 4.85,
 25: 4.81,
 30: 4.74,
 523: 4.69,
 348: 4.67,
 171: 4.63,
 452: 4.56,
 43: 4.55,
 122: 4.55,
 371: 4.55,
 441: 4.52,
 400: 4.51,
 52: 4.48,
 538: 4.47,
 168: 4.46,
 417: 4.46,
 543: 4.45,
 106: 4.44,
 319: 4.43,
 601: 4.43,
 413: 4.41,
 475: 4.41,
 188: 4.4,
 12: 4.39,
 276: 4.39,
 154: 4.38,
 533: 4.38,
 581: 4.38,
 1: 4.37,
 69: 4.37,
 586: 4.37,
 59: 4.36,
 128: 4.36,
 544: 4.36,
 253: 4.35,
 459: 4.35,
 553: 4.34,
 585: 4.34,
 519: 4.33,
 70: 4.32,
 336: 4.32,
 435: 4.32,
 300: 4.3,
 93: 4.29,
 532: 4.28,
 74: 4.27,
 49: 4.26,
 80: 4.26,
 224: 4.26,
 291: 4.26,
 164: 4.25,
 169: 4.25,
 246: 4.25,
 398: 4.25,
 162: 4.24,
 209: 4.24,
 362: 4.24,
 494: 4.23,
 250: 4.22,
 17: 4.21,
 337: 4.21,
 340: 4.21,
 573: 4.21,
 79: 4.2,
 227: 4.2,
 252: 4.2,
 258: 4.2,
 460: 4.2,
 595: 4.2,
 97: 4.19,
 364: 4.19,
 526: 4.19,
 119: 4.18,
 267: 4.18,
 505: 4.18,
 296: 4.17,
 72: 4.16,
 465: 4.16,
 491: 4.16,
 458: 4.15,
 29: 4.14,
 37: 4.14,
 243: 4.14,
 290: 4.14

### 💁 Дисперсия самого противоречивого пользователя - 4.37

А вот самый противоречивый пользователь (3, дисперсия 4.37) - это уже не случайность двух оценок: у него 39 оценок, то есть он действительно оценивает фильмы то очень высоко, то очень низко на приличной выборке. В отличие от «спорных фильмов» выше, здесь высокая дисперсия - устойчивая характеристика поведения, а не артефакт малой выборки.

In [36]:
%timeit ratings_users.top_controversial_users(5)
ratings_users.top_controversial_users(5)

6.48 ms ± 96.7 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


{3: 4.37, 55: 3.22, 461: 3.22, 259: 3.05, 329: 3.05}

## **V. Обогащение через TMDB: бюджеты, сборы, режиссёры**

**Для ускорения работы с TMDB API анализ в этом разделе выполнен на подвыборке из 30 фильмов - первых по movieId в датасете. Поскольку movieId в MovieLens в целом коррелирует с хронологией добавления фильмов в каталог, эта выборка смещена в сторону более старых релизов и не является случайной или репрезентативной для всего датасета.*

*Логика подсчёта и сам код работают корректно и не требуют изменений - ограничение носит исключительно характер выборки данных, а не ошибку в реализации.*

***Приведённые ниже цифры стоит рассматривать как иллюстрацию работы пайплайна, а не как итоговые выводы о полном датасете MovieLens.***

In [37]:
%timeit Links(LINKS_PATH, MOVIES_PATH, TMDB_API_KEY)
links = Links(LINKS_PATH, MOVIES_PATH, TMDB_API_KEY)
links.movies_data

32.9 ms ± 9.71 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


{1: {'tmdbId': 862, 'title': 'Toy Story (1995)'},
 2: {'tmdbId': 8844, 'title': 'Jumanji (1995)'},
 3: {'tmdbId': 15602, 'title': 'Grumpier Old Men (1995)'},
 4: {'tmdbId': 31357, 'title': 'Waiting to Exhale (1995)'},
 5: {'tmdbId': 11862, 'title': 'Father of the Bride Part II (1995)'},
 6: {'tmdbId': 949, 'title': 'Heat (1995)'},
 7: {'tmdbId': 11860, 'title': 'Sabrina (1995)'},
 8: {'tmdbId': 45325, 'title': 'Tom and Huck (1995)'},
 9: {'tmdbId': 9091, 'title': 'Sudden Death (1995)'},
 10: {'tmdbId': 710, 'title': 'GoldenEye (1995)'},
 11: {'tmdbId': 9087, 'title': 'American President, The (1995)'},
 12: {'tmdbId': 12110, 'title': 'Dracula: Dead and Loving It (1995)'},
 13: {'tmdbId': 21032, 'title': 'Balto (1995)'},
 14: {'tmdbId': 10858, 'title': 'Nixon (1995)'},
 15: {'tmdbId': 1408, 'title': 'Cutthroat Island (1995)'},
 16: {'tmdbId': 524, 'title': 'Casino (1995)'},
 17: {'tmdbId': 4584, 'title': 'Sense and Sensibility (1995)'},
 18: {'tmdbId': 5, 'title': 'Four Rooms (1995)'},
 

In [74]:
links = Links("../datasets/links.csv", "../datasets/movies.csv", TMDB_API_KEY)

top_30_ids = set(list(links.movies_data.keys())[:30])

links.movies_data = {
    movie_id: info
    for movie_id, info in links.movies_data.items()
    if movie_id in top_30_ids
}

### Общая информация о фильме

Для заданного списка movieId метод тянет данные из TMDB и собирает список по запрошенным полям. Сортировка по movieId по убыванию.

In [67]:
%timeit links.get_imdb([1, 2, 3], ["Title", "Director", "Budget", "Runtime"])
links.get_imdb([1, 2, 3], ["Title", "Director", "Budget", "Runtime"])

1.03 s ± 20.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


[[3, 'Grumpier Old Men', 'Howard Deutch', 25000000, 101],
 [2, 'Jumanji', 'Joe Johnston', 65000000, 104],
 [1, 'Toy Story', 'John Lasseter', 30000000, 81]]

### Наиболее дорогие фильмы

Возвращает top-n {название: бюджет} по всем фильмам в self.movies_data, отсортированных по убыванию бюджета.

In [75]:
links.most_expensive(3)

{'Cutthroat Island': 102000000, 'Money Train': 68000000, 'Jumanji': 65000000}

### Наиболее продуктивные режиссёры

Проходит по всем фильмам в self.movies_data, считает, у скольких фильмов каждый режиссёр значится как Director, возвращает top-n {режиссёр: количество фильмов}.

In [69]:
%timeit links.top_directors(3)
links.top_directors(3)

12.3 s ± 362 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


{'John Lasseter': 1, 'Joe Johnston': 1, 'Howard Deutch': 1}

### Фильмы с самой дорогой минутой

Возвращает top-n {название: бюджет / хронометраж}, отсортировано по убыванию - учитываются только фильмы с ненулевыми бюджетом и хронометражом.

In [70]:
%timeit links.top_cost_per_minute(3)
links.top_cost_per_minute(3)

12.1 s ± 1.13 s per loop (mean ± std. dev. of 7 runs, 1 loop each)


{'Cutthroat Island': 822580.65, 'Jumanji': 625000.0, 'Money Train': 618181.82}

### Наиболее длинные фильмы

Возвращает top-n {название: хронометраж} по убыванию длительности.

In [71]:
%timeit links.longest(3)
links.longest(3)

11.9 s ± 576 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


{'Nixon': 192, 'Casino': 179, 'Heat': 170}

### Самые окупаемые фильмы

Возвращает top-n {название: прибыль}, где прибыль = выручка - бюджет, отсортировано по убыванию.

In [72]:
%timeit links.most_profitable(3)
links.most_profitable(3)

11 s ± 978 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


{'Toy Story': 932301978, 'GoldenEye': 292194034, 'Jumanji': 197821940}